x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,180 @@
-
-	// // If superfluid delegated, superfluid undelegate
-	// // This also burns the underlying synthetic osmo
-	// err = k.unbondSuperfluidIfExists(ctx, sender, lockId)
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-
-	// // Finish unlocking directly for locked locks
-	// // this also unlocks locks that were in the unlocking queue
-	// err = k.lk.ForceUnlock(ctx, *lock)
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-
-	// // If shares to migrate is not specified, we migrate all shares.
-	// if sharesToMigrate.IsZero() {
-	// 	sharesToMigrate = gammSharesInLock
-	// }
-
-	// // Otherwise, we must ensure that the shares to migrate is less than or equal to the shares in the lock.
-	// if sharesToMigrate.Amount.GT(gammSharesInLock.Amount) {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
-	// }
-
-	// // Exit the balancer pool position.
-	// exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-	// // Defense in depth, ensuring we are returning exactly two coins.
-	// if len(exitCoins) != 2 {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
-	// }
-
-	// // Create a full range (min to max tick) concentrated liquidity position.
-	// positionId, amount0, amount1, liquidity, joinTime, err = k.clk.CreateFullRangePosition(ctx, concentratedPool, sender, exitCoins)
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-
-	// // If there are remaining gamm shares, we must re-lock them.
-	// remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
-	// if !remainingGammShares.IsZero() {
-	// 	newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
-	// 	newLockId = newLock.ID
-	// 	if err != nil {
-	// 		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 	}
-	// 	// If the lock was unlocking, we begin the unlock from where it left off.
-	// 	if preUnlockLock.IsUnlocking() {
-	// 		_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
-	// 		if err != nil {
-	// 			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 		}
-	// 	}
-	// }
-
-	return positionId, amount0, amount1, liquidity, joinTime, poolIdLeaving, poolIdEntering, newLockId, nil
+func (k Keeper) MigrateLockedPositionFromBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, poolIdLeaving, poolIdEntering, gammLockId, concentratedLockId uint64, err error) {
```

Comment:
nit: consider refactoring this method into a method that calls sub-methods depending on whether it is superfluid delegated / undelegating. The sub methods would each handle migrating superfluid delegated lock, migrating undelegating superfluid lock. I think this way we can have better code sanity instead of doing if statements throughout the method

Reply:
Agreed, addressed this here https://github.com/osmosis-labs/osmosis/pull/4885/commits/2410b7e1b1fa07e10c31ea646edf6f84ba9b05c0

Made a single entry function that calls one of three separate functions depending on if it is sfDelegated, sfUndelegating, or just locked/unlocking